### PR TITLE
refactor: update types

### LINKED ISSUE #68: Error `TS(2589): Type instantiation is excessively deep and possibly infinite`
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.25.1

### Plugin version

2.2.2

### Node.js version

20.10

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

22.04.3 LTS

### Description

I get error `TS(2589): Type instantiation is excessively deep and possibly infinite` whenever relying solely on this plugin. I have to manually create an interface or type per request(which defeats the whole purpose)

### Steps to Reproduce

```ts
export const SampleController: FastifyPluginAsyncJsonSchemaToTs = async (fastify, _opts) => {
  fastify.post(
  "/test",
  {
    schema: {
      body: {
        type: "object",
        properties: {
                dateFrom: { type: "string", format: "date-time" },
        },
        required: ["dateFrom"],
      },
      response: {
        200: {
          type: "object",
          properties: {
            message: { type: "string" },
          },
        },
        400: {
          type: "object",
          properties: {
            error: { type: "string" },
            message: { type: "string" },
          },
        },
      },
    },
  },
  async (request, reply): Promise<void> => {
	  const {dateFrom: dateString} = request.body; // Error
  };
```

But if I do 
```ts
fastify.post<{ Body: { dateFrom: string } }>(...)
```

it works just fine. Issue is I don't want to have to manually define schemas and types, hence why I use this plugin.

### Expected Behavior

I would expect the types to be successfully generated from the schema, with no manual definition from my side

### PR DESCRIPTION
**Pull request by climba03003**
Closes #70 Closes #71
Fixes #68 

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by climba03003** [comment]
It somehow working on my side. 
Keep all packages the latest always the key in TypeScript world.

**Comment by climba03003** [comment]
Interesting, it error only when using `tsc` but not the runtime.

**Comment by mcollina** [comment]
> The problem is non-strict mode.

Even with this change, tsserver still complains on my machine.

**Comment by climba03003** [comment]
> Even with this change, tsserver still complains on my machine.

The same error? Maybe try it from fresh.

```shell
git clone https://github.com/fastify/fastify-type-provider-json-schema-to-ts.git
cd fastify-type-provider-json-schema-to-ts
git checkout update
npm install
npm run build
npm run typescript
```

**Comment by mcollina** [comment]
The compilation still fail without `--strict` as well.

**Comment by jfet97** [comment]
The `strictNullChecks` flag seems to be enough: ` npx tsc --strictNullChecks  --noEmit types/*.ts` doesn't emit any error.

Probably the library you are using relies on having `null` and `undefined` as distinct types while checking some base case for some recursive computation.

If `strictNullChecks` is `false`, `null` and `undefined` are assignable to all types, therefore any `extends` checks could change behaviour if the checked type is  `null` or `undefined`.

**Comment by jfet97** [comment]
The documentation of json-schema-to-ts says exactly that `strict: true` is required

**Comment by Uzlopak** [comment]
@jfet97 
Good point. If that is so, we should document it also in this repo.

File: README.md [hunk 1]
```diff
@@ -1,13 +1,30 @@
...
-A Type Provider for json-schema-to-ts
+A Type Provider for [json-schema-to-ts](https://github.com/ThomasAribart/json-schema-to-ts)
 
 ## Install
 
 ```
 npm i @fastify/type-provider-json-schema-to-ts
 ```
 
+## TypeScript requirements
```
Comment (climba03003) [inline README.md:11]: @mcollina I updated the requirements based on the information provided by @jfet97

File: README.md [hunk 2]
```diff
@@ -1,13 +1,30 @@
...
+
+It is required to use `TypeScript@4.3` or above with
+[`strict`](https://www.typescriptlang.org/tsconfig#strict)
+mode enabled and
+[`noStrictGenericChecks`](https://www.typescriptlang.org/tsconfig#noStrictGenericChecks)
+disabled. You may take the below configuration as an example.
```
Comment (Uzlopak) [inline README.md:17]: Maybe make it dummy proof and say that this is a tsconfig.json...

Comment (Uzlopak) [reply README.md]: Thanks ;)

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**